File: python/npc_maker/indiv.py
"""
Data structure to represent an individual life-form
"""

from pathlib import Path
from .utils import clean_command
import io
import json
import math
import os
import tempfile
import uuid

__all__ = (
    "Individual",
)

def _check_genome(genome):
    assert type(genome) is bytes
    assert len(genome) > 0

# All individuals have these fields, with default value constructors.
_standard_fields = {
    "name": None,
    "environment": None,
    "body_type": None,
    "controller": None,
    "score": type(None),
    "telemetry": dict,
    "epigenome": dict,
    "species": lambda: str(uuid.uuid4()),
    "parents": list,
    "children": list,
    "generation": int,
    "ascension": type(None),
    "birth_date": str,
    "death_date": str,
}

class Individual:
    """
    Container for a distinct life-form and all of its associated data
    """
    def __init__(self,
                environment: str,
                body_type: str,
                controller: [str],
                genome: bytes):
        """
        Create a new individual. This is used to initialize new populations
        """
        _check_genome(genome)
        self.name           = str(uuid.uuid4())
        self.environment    = str(environment)
        self.body_type      = str(body_type)
        # Don't resolve the path because the PWD could change.
        self.controller     = clean_command(controller, resolve=False)
        self.genome         = genome
        self.score          = None
        self.telemetry      = {}
        self.epigenome      = {}
        self.species        = str(uuid.uuid4())
        self.parents        = []
        self.children       = []
        self.birth_date     = ""
        self.death_date     = ""
        self.generation     = 0
        self.ascension      = None
        self.extra          = {}
        self.path           = None

    def get_name(self) -> str:
        """
        Get this individual's name, which is a UUID string
        """
        return self.name

    def get_environment(self) -> str:
        """
        Get the name of environment which contains this individual
        """
        return self.environment

    def get_body_type(self) -> str:
        """
        Get the name of this individual's body_type
        """
        return self.body_type

    def get_controller(self) -> list:
        """
        Get the command line invocation for the controller program
        """
        return list(self.controller)

    def get_genome(self) -> bytes:
        """
        Get this individual's genetic data,
        which is an immutable byte array
        """
        if self.genome is None:
            with open(self.path, 'rb') as file:
                data = file.read()
            metadata, self.genome = data.split(b'\x00', maxsplit=1)
        return self.genome

    def get_score(self) -> str:
        """
        Get the most recently assigned score,
        or None if it has not been assigned yet
        """
        return self.score

    def get_custom_score(self, score_function="score") -> float:
        """
        Apply a custom scoring function to this individual

        Several classes accept an optional custom score function,
        and they delegate to this method.

        Argument score_function must be one of the following:
            * A callable function: f(individual) -> float,
            * The word "score",
            * The word "ascension",
            * A key in the individual's telemetry dictionary. The corresponding
              value will be converted into a float.

        A score of None is converted into negative infinity.
        """
        if callable(score_function):
            score = score_function(self)
        elif not score_function or score_function == "score":
            score = self.score
        elif score_function == "ascension":
            score = self.ascension
        elif score_function in self.telemetry:
            score = self.telemetry[score_function]
        else:
            raise ValueError("unrecognized score function " + repr(score_function))
        # 
        if score is None:
            score = -math.inf
        # 
        return float(score)

    def get_telemetry(self) -> dict:
        """
        Get the environmental info dictionary

        Returns a reference to the individual's internal "telemetry" dictionary;
        modifications are permanent.
        """
        return self.telemetry

    def get_epigenome(self) -> dict:
        """
        Get the epigenetic info dictionary

        Returns a reference to the individual's internal "epigenome" dictionary;
        modifications are permanent.
        """
        return self.epigenome

    def get_species(self) -> str:
        """
        Get the species UUID

        Mating may be restricted to individuals of the same species.
        """
        return self.species

    def get_parents(self) -> [str]:
        """
        Get the names of this individual's parents
        """
        return list(self.parents)

    def get_children(self) -> [str]:
        """
        Get the names of this individual's children
        """
        return list(self.children)

    def get_birth_date(self) -> str:
        """
        The time of birth, as a UTC timestamp,
        or an empty string if this individual has not yet been born
        """
        return self.birth_date

    def get_death_date(self) -> str:
        """
        The time of death, as a UTC timestamp,
        or an empty string if this individual has not yet died
        """
        return self.death_date

    def get_generation(self) -> int:
        """
        How many cohorts of the population size passed before this individual was born?
        """
        return self.generation

    def get_ascension(self) -> int:
        """
        How many individuals died before this individual?
        Returns None if this individual has not yet died.
        """
        return self.ascension

    def get_extra(self) -> dict:
        """
        Get all custom / unofficial fields that are saved with the individual

        Returns a reference to this individual's internal data.
        Changes made to the returned value will persist with the individual.
        """
        return self.extra

    def get_path(self) -> Path:
        """
        Returns the file path this individual was loaded from or saved to.
        Returns None if this individual has not touched the file system.
        """
        return self.path

    def asex(self, child_genome: bytes) -> "Individual":
        """
        Asexually reproduce an individual
        """
        _check_genome(child_genome)
        cls = type(self)
        child = cls(child_genome,
                environment = self.environment,
                body_type   = self.body_type,
                controller  = self.controller,
                epigenome   = self.epigenome,
                species     = self.species,
                generation  = self.generation + 1,
                parents     = [self.name])
        self.children.append(child.name)
        return child

    @classmethod
    def sex(cls, parents: ["Individual"], child_genome: bytes) -> "Individual":
        """
        Sexually reproduce the given individuals
        """
        parents = list(parents)
        # Technically the spec requires 2 parents, 1 parent should not crash it
        assert len(parents) >= 1
        assert all(isinstance(p, cls) for p in parents)
        _check_genome(child_genome)
        self = parents[0]
        child = cls(child_genome,
                environment = self.environment,
                body_type   = self.body_type,
                species     = self.species,
                controller  = self.controller,
                generation  = max(p.generation for p in parents) + 1,
                parents     = list(p.name for p in parents))
        # Update the parent's children.
        for p in parents:
            p.children.append(child.name)
        return child

    def save(self, path=None) -> Path:
        """
        Serialize this individual to JSON and write it to a file

        Argument path is the directory to save in. Optional, if missing will
        either overwrite the previous save file or save to temporary directory.
        The filename will be the individual's name with the ".indiv" file extension.

        Returns the file path of the saved individual.
        """
        if not path:
            if self.path:
                path = self.path.parent
            else:
                path = tempfile.gettempdir()
        path = Path(path)
        # Make the directory in case this is the first individual to be saved to it.
        if not path.exists():
            path.mkdir()
        assert path.is_dir()
        path = path.joinpath(self.name + ".indiv")
        # Load genome from file before modifying the file system.
        genome = self.get_genome()
        # Unofficial fields, in case of conflict these take lower precedence.
        data = dict(self.extra)
        # Official fields, as described by the specification.
        for attribute in _standard_fields:
            value = getattr(self, attribute, None)
            if value:
                data[attribute] = value
        # Convert paths to strings for JSON serialization.
        if self.controller:
            data["controller"]    = list(data["controller"])
            data["controller"][0] = str(data["controller"][0])
        # 
        data = json.dumps(data)
        # Save to a temporary file, sync, and atomic move into place.
        fd, tmp_path = tempfile.mkstemp()
        file = os.fdopen(fd, "wb")
        file.write(data.encode("utf-8"))
        file.write(b'\x00')
        file.write(genome)
        file.flush()
        file.close()
        Path(tmp_path).rename(path)
        self.path = path
        return path

    @classmethod
    def load(cls, path) -> 'Individual':
        """
        Load a previously saved individual

        Returns None if the given path does not end with ".indiv"
        """
        path = Path(path)
        if path.suffix.lower() != ".indiv":
            return
        text = b''
        with open(path, 'rb') as file:
            while True:
                chunk = file.read(io.DEFAULT_BUFFER_SIZE)
                split = chunk.split(b'\x00', maxsplit=1)
                text += split[0]
                if len(split) > 1:
                    break
        metadata = json.loads(text)
        self = cls.__new__(cls, **metadata)
        for attribute, default_value in _standard_fields.items():
            if attribute in metadata:
                value = metadata.pop(attribute)
            elif attribute == "name":
                value = path.stem
            elif default_value is not None:
                value = default_value()
            else:
                continue
            setattr(self, attribute, value)
        self.extra = metadata
        self.path = path
        self.genome = None
        # Convert controller path back to Path objects.
        if getattr(self, "controller", False):
            self.controller[0] = Path(self.controller[0])
        return self

File: rust/src/indiv.rs
//! Data structure to represent an individual life-form

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader, BufWriter, Read, Result, Write};
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex, OnceLock};

/// Generate a universally unique name. This will never return the same name twice.
fn uuid4() -> String {
    let uuid: u128 = rand::random();
    format!("{uuid:032X}")
}

/// Container for a distinct life-form and all of its associated data
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Individual {
    /// Name or UUID of this individual
    #[serde(default)]
    pub name: String,

    /// Number of individuals who died before this one,
    /// or None if this individual has not yet died
    #[serde(default)]
    pub ascension: Option<u64>,

    /// Name of the environment that this individual lives in
    #[serde(default)]
    pub environment: String,

    /// Name of the body_type that this individual uses
    #[serde(default)]
    pub body_type: String,

    /// Name or UUID of this individual's species.
    /// Mating may be restricted to individuals of the same species.
    #[serde(default)]
    pub species: String,

    /// Command line invocation of the controller program
    #[serde(default)]
    pub controller: Vec<String>,

    /// Genetic parameters for this individual
    #[serde(skip)]
    pub genome: OnceLock<Arc<[u8]>>,

    /// The environmental info dictionary. The environment updates this information.
    #[serde(default)]
    pub telemetry: HashMap<String, String>,

    /// The epigenetic info dictionary. The controller updates this information.
    #[serde(default)]
    pub epigenome: HashMap<String, String>,

    /// Reproductive fitness of this individual, as assessed by the environment
    #[serde(default)]
    pub score: Option<String>,

    /// Number of cohorts that passed before this individual was born
    #[serde(default)]
    pub generation: u64,

    /// The names of this individual's parents
    #[serde(default)]
    pub parents: Vec<String>,

    /// The names of this individual's children
    #[serde(default)]
    pub children: Vec<String>,

    /// Time of birth, as a UTC timestamp, or an empty string if this individual
    /// has not yet been born
    #[serde(default)]
    pub birth_date: String,

    /// Time of death, as a UTC timestamp, or an empty string if this individual
    /// has not yet died
    #[serde(default)]
    pub death_date: String,

    /// Custom / unofficial fields that are saved with the individual
    #[serde(flatten)]
    pub extra: HashMap<String, serde_json::Value>,

    /// The file path this individual was loaded from or saved to, or None
    /// if this individual has not touched the file system
    #[serde(skip)]
    pub path: Option<PathBuf>,
}

impl Individual {
    /// Create a new individual. This is used to initialize new populations
    pub fn new(environment: &str, body_type: &str, controller: &[&str], genome: Box<[u8]>) -> Individual {
        assert!(!controller.is_empty());
        assert!(!genome.is_empty());
        Individual {
            name: uuid4(),
            ascension: None,
            environment: environment.to_string(),
            body_type: body_type.to_string(),
            species: uuid4(),
            controller: controller.iter().map(|arg| arg.to_string()).collect(),
            genome: OnceLock::from(Arc::from(genome)),
            telemetry: HashMap::new(),
            epigenome: HashMap::new(),
            score: None,
            generation: 0,
            parents: Vec::new(),
            children: Vec::new(),
            birth_date: String::new(),
            death_date: String::new(),
            extra: HashMap::new(),
            path: None,
        }
    }

    /// Get the genetic parameters for this individual. This loads the genome
    /// from file if necessary.
    pub fn genome(&self) -> Result<Arc<[u8]>> {
        Ok(self
            .genome
            .get_or_init(|| {
                // Unwrap everything because "get_or_try_init" is unstable.
                let path = self.path.as_ref().expect("missing genome");
                let mut file = BufReader::new(File::open(path).unwrap());
                file.skip_until(b'\0').expect("missing genome");
                let mut data = vec![];
                file.read_to_end(&mut data).unwrap();
                Arc::from(data)
            })
            .clone())
    }

    /// Asexually reproduce an individual
    pub fn asex(&mut self, child_genome: &[u8]) -> Individual {
        assert!(!child_genome.is_empty());
        let individual = Individual {
            name: uuid4(),
            ascension: None,
            environment: self.environment.clone(),
            body_type: self.body_type.clone(),
            species: self.species.clone(),
            controller: self.controller.clone(),
            genome: OnceLock::from(Arc::from(child_genome)),
            telemetry: HashMap::new(),
            epigenome: HashMap::new(),
            score: None,
            generation: self.generation + 1,
            parents: vec![self.name.clone()],
            children: vec![],
            birth_date: String::new(),
            death_date: String::new(),
            extra: HashMap::new(),
            path: None,
        };
        self.children.push(individual.name.clone());
        individual
    }

    // TODO: Sex allows more than two parents, and should not crash with only one parent.

    /// Sexually reproduce two individuals
    pub fn sex(&mut self, other: &mut Individual, child_genome: &[u8]) -> Individual {
        assert!(!child_genome.is_empty());
        let individual = Individual {
            name: uuid4(),
            ascension: None,
            environment: self.environment.clone(),
            body_type: self.body_type.clone(),
            species: self.species.clone(),
            controller: self.controller.clone(),
            genome: OnceLock::from(Arc::from(child_genome)),
            telemetry: HashMap::new(),
            epigenome: HashMap::new(),
            score: None,
            generation: self.generation.max(other.generation) + 1,
            parents: vec![self.name.clone(), other.name.clone()],
            children: vec![],
            birth_date: String::new(),
            death_date: String::new(),
            extra: HashMap::new(),
            path: None,
        };
        self.children.push(individual.name.clone());
        other.children.push(individual.name.clone());
        individual
    }

    fn file_name(&self) -> String {
        format!("{}.indiv", self.name)
    }

    /// Save an individual to a file
    ///
    /// Argument path is the directory to save in. Optional, use empty string
    /// to either overwrite the previous save-file or to use a temporary directory.
    /// The filename will be the individual's name with the ".indiv" file extension.
    ///
    /// Returns the file path of the saved individual.
    pub fn save(&mut self, path: impl AsRef<Path>) -> Result<&Path> {
        let mut path: PathBuf = path.as_ref().into();
        // Fill in default path.
        if path.to_str() == Some("") {
            if let Some(save_file) = self.path.as_ref() {
                path = save_file.parent().unwrap().into();
            } else {
                path = std::env::temp_dir();
            }
        }
        // Load genome from file before modifying the file system.
        self.genome()?;
        // Make the directory in case this is the first individual to be saved to it.
        if !path.exists() {
            std::fs::create_dir(&path)?;
        }
        // Make paths to temporary buffer and final file locations.
        let file_name = self.file_name();
        let mut temp = std::env::temp_dir();
        temp.push(format!("{file_name}.tmp"));
        path.push(file_name);
        //
        let file = File::create(&temp)?;
        let mut buf = BufWriter::new(file);
        serde_json::to_writer(&mut buf, self).unwrap();
        buf.write_all(b"\0")?;
        buf.write_all(&self.genome()?)?;
        let file = buf.into_inner()?; // flush the buffer
        file.sync_all()?; // push to disk
        std::fs::rename(&temp, &path)?; // move file into place
        self.path = Some(path);
        self.genome.take(); // free the genome
        Ok(self.path.as_ref().unwrap())
    }

    /// Load a previously saved individual
    ///
    /// The file extension must be ".indiv"
    pub fn load(path: impl AsRef<Path>) -> Result<Individual> {
        let path = path.as_ref().to_path_buf();
        // assert!(path.get_ext() == ".indiv"); // TODO
        let mut file = BufReader::new(File::open(&path)?);
        let mut text = vec![];
        file.read_until(b'\0', &mut text)?;
        text.pop_if(|&mut char| char == b'\0');
        let mut individual: Individual = serde_json::from_slice(&text)?;
        individual.path = Some(path);
        Ok(individual)
    }

    /// Get all individuals saved in a directory
    pub fn load_dir(path: impl AsRef<Path>) -> Result<Vec<Individual>> {
        let path = path.as_ref();
        if !path.exists() {
            return Ok(vec![]);
        }
        let mut individuals = vec![];
        for file in path.read_dir()? {
            let file = file?;
            if !file.file_type()?.is_file() {
                continue;
            }
            let file_name = file.file_name();
            let Some(file_name) = file_name.to_str() else {
                continue;
            };
            if file_name.ends_with(".indiv") {
                individuals.push(Individual::load(file.path())?);
            }
        }
        Ok(individuals)
    }

    /// Remove this individual and its associated save file
    pub fn delete(self) -> Result<()> {
        if let Some(path) = self.path {
            std::fs::remove_file(path)?;
        }
        Ok(())
    }

    /// Delete the individual if this is the last reference to it
    pub fn drop(this: Arc<Mutex<Self>>) -> Result<()> {
        if let Some(mutex) = Arc::into_inner(this) {
            let individual = mutex.into_inner().unwrap();
            individual.delete()?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn uuid4_len() {
        for _ in 0..100 {
            assert_eq!(uuid4().len(), 32);
        }
    }
    #[test]
    fn uuid4_unique() {
        use std::collections::HashSet;
        let unique = 1000;
        assert_eq!((0..unique).map(|_| uuid4()).collect::<HashSet<String>>().len(), unique);
    }
    #[test]
    fn indiv_save_load() {
        let mut a = Individual::new("env", "body", &["ctrl"], Box::new(*b"genome"));
        let path = a.save("").unwrap();
        let b = Individual::load(path).unwrap();
        assert_eq!(a, b);
        b.delete().unwrap();
    }
}
